pkg/pillar/cmd/zedmanager/zedmanager.go

```diff
@@ -768,137 +770,174 @@ func handleDelete(ctx *zedmanagerContext, key string,
 	log.Infof("handleDelete done for %s", status.DisplayName)
 }
 
-// Returns needRestart, needPurge
+// Returns needRestart, needPurge, plus a string for each.
 // If there is a change to the disks, adapters, or network interfaces
 // it returns needPurge.
 // If there is a change to the CPU etc resources it returns needRestart
 // Changes to ACLs don't result in either being returned.
 func quantifyChanges(config types.AppInstanceConfig,
-	status types.AppInstanceStatus) (bool, bool) {
+	status types.AppInstanceStatus) (bool, bool, string, string) {
 
 	needPurge := false
 	needRestart := false
+	var purgeReason, restartReason string
 	log.Infof("quantifyChanges for %s %s",
 		config.Key(), config.DisplayName)
 	if len(status.StorageStatusList) != len(config.StorageConfigList) {
-		log.Infof("quantifyChanges len storage changed from %d to %d",
+		str := fmt.Sprintf("number of volumes changed from %d to %d",
 			len(status.StorageStatusList),
 			len(config.StorageConfigList))
+		log.Infof(str)
 		needPurge = true
+		purgeReason += str + "\n"
 	} else {
 		for _, sc := range config.StorageConfigList {
 			ss := lookupStorageStatus(&status, sc)
 			if ss == nil {
-				log.Errorf("quantifyChanges missing StorageStatus for (Name: %s, "+
+				str := fmt.Sprintf("missing StorageStatus for (Name: %s, "+
 					"ImageSha256: %s, ImageID: %s, PurgeCounter: %d)",
 					sc.Name, sc.ImageSha256, sc.ImageID, sc.PurgeCounter)
+				log.Errorf(str)
 				needPurge = true
+				purgeReason += str + "\n"
 				continue
 			}
 			if ss.ImageID != sc.ImageID {
-				log.Infof("quantifyChanges storage imageID changed from %s to %s",
+				str := fmt.Sprintf("storage imageID changed from %s to %s",
 					ss.ImageID, sc.ImageID)
+				log.Infof(str)
 				needPurge = true
+				purgeReason += str + "\n"
 			}
 			if ss.ReadOnly != sc.ReadOnly {
-				log.Infof("quantifyChanges storage ReadOnly changed from %v to %v",
-					ss.ReadOnly, sc.ReadOnly)
+				str := fmt.Sprintf("storage ReadOnly changed from %v to %v for %s",
+					ss.ReadOnly, sc.ReadOnly, ss.ImageID)
```

Comment:
Isn't it better to include storage status name instead of image ID in purge reason?

Reply:
Or both ...
But this will go away when the app instance uses volumeRef instead of Drive, so I didn't bother figuring out the best way to identify it; when we use volumeRef we will just check for the number of volumes and whether or not the volumeRef uuid/generationCounter has changed.